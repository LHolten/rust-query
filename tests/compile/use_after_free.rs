use rust_query::{Database, IntoExpr, LocalClient, Table, migration::schema};

#[schema(Schema)]
pub mod vN {
    pub struct MyTable {
        pub name: String,
    }
}
use v0::*;

fn test(db: Database<Schema>) {
    let mut client = LocalClient::try_new().unwrap();

    let txn = client.transaction(&db);
    let items = txn.query(|rows| {
        let item = MyTable::join(rows);
        rows.into_vec(item)
    });
    let items: Vec<_> = items.into_iter().map(|x| x.into_expr()).collect();
    drop(txn);

    let txn = client.transaction(&db);
    for item in items {
        let name = txn.query_one(item.name());
        println!("{name}")
    }
}

fn main() {}
