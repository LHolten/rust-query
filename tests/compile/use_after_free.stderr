error[E0499]: cannot borrow `client` as mutable more than once at a time
  --> tests/compile/use_after_free.rs:22:15
   |
14 |     let txn = client.transaction(&db);
   |               ------ first mutable borrow occurs here
...
22 |     let txn = client.transaction(&db);
   |               ^^^^^^ second mutable borrow occurs here
23 |     for item in items {
   |                 ----- first borrow later used here
